plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -15,89 +15,165 @@
-        }
-        catch (IOException e) {
-            throw new UncheckedIOException(e);
+        checkLoaderException();
+        loaderTask.close();
+    }
+
+    private void onLoaderFailure(Throwable exception)
+    {
+        loaderException = exception;
+        loaderTask.close();
+        splitQueue.finish();
+    }
+
+    private void checkLoaderException()
+    {
+        if (loaderException != null) {
+            if (loaderException instanceof TrinoException) {
+                throw (TrinoException) loaderException;
+            }
+            throw new TrinoException(GENERIC_INTERNAL_ERROR, loaderException);
         }
     }
 
-    private ConnectorSplit toIcebergSplit(FileScanTask task)
+    private static class IcebergSplitLoader
+            implements ResumableTask
     {
-        return new IcebergSplit(
-                task.file().path().toString(),
-                task.start(),
-                task.length(),
-                task.file().fileSizeInBytes(),
-                task.file().format(),
-                ImmutableList.of(),
-                getPartitionKeys(task));
+        private final SchemaTableName schemaTableName;
+        private final AsyncQueue<ConnectorSplit> splitQueue;
+        private final CloseableIterable<CombinedScanTask> combinedScanIterable;
+        private final Iterator<FileScanTask> fileScanIterator;
+
+        public IcebergSplitLoader(
+                SchemaTableName schemaTableName,
+                TableScan tableScan,
+                AsyncQueue<ConnectorSplit> splitQueue)
+        {
+            this.schemaTableName = requireNonNull(schemaTableName, "schemaTableName is null");
+            this.splitQueue = requireNonNull(splitQueue, "splitQueue is null");
+
+            this.combinedScanIterable = tableScan.planTasks();
+            this.fileScanIterator = Streams.stream(combinedScanIterable)
+                    .map(CombinedScanTask::files)
+                    .flatMap(Collection::stream)
+                    .iterator();
+        }
+
+        @Override
+        public TaskStatus process()
+        {
+            while (fileScanIterator.hasNext()) {
+                FileScanTask task = fileScanIterator.next();
+                if (!task.deletes().isEmpty()) {
+                    throw new TrinoException(NOT_SUPPORTED, "Iceberg tables with delete files are not supported: " + schemaTableName);
+                }
+                IcebergSplit icebergSplit = toIcebergSplit(task);
+                ListenableFuture<Void> queueOfferFuture = splitQueue.offer(icebergSplit);
+                if (!queueOfferFuture.isDone()) {
+                    // Queue is full, wait until there is space
+                    return TaskStatus.continueOn(queueOfferFuture);
+                }
+            }
+            splitQueue.finish();
+            return TaskStatus.finished();
+        }
+
+        public void close()
+        {
+            if (combinedScanIterable != null) {
```

Comment:
can `cSI` be null? It does not seem so. 
Replace with verify? Or just ignore nullability?